map/client/client.go

```diff
@@ -0,0 +1,335 @@
+	return r.c.Put(i, k.Sign(payload), payload)
+}
+
+// PayloadSize implements RWClient.PayloadSize
+func (r *rwMap) PayloadSize() int {
+	return r.c.PayloadSize()
+}
+
+// roMap implements ReadOnlyTranport using a ReadOnlyCap and reference to map client
+type roMap struct {
+	c     *Client
+	roCap common.ReadOnlyCap
+}
+
+// Get implements ROClient.Get
+func (r *roMap) Get(addr []byte) ([]byte, error) {
+	i := r.roCap.Addr(addr)
+	k := r.roCap.Read(addr)
+	return r.c.Get(i, k.Sign(i.Bytes()))
+}
+
+// PayloadSize implements ROClient.PayloadSize
+func (r *roMap) PayloadSize() int {
+	return r.c.PayloadSize()
+}
+
+// woMap implements WOClient
+type woMap struct {
+	c     *Client
+	woCap common.WriteOnlyCap
+}
+
+// Put implements WOClient.Put
+func (w *woMap) Put(addr []byte, payload []byte) error {
+	i := w.woCap.Addr(addr)
+	k := w.woCap.Write(addr)
+	return w.c.Put(i, k.Sign(payload), payload)
+}
+
+// PayloadSize implements WOClient.PayloadSize
+func (r *woMap) PayloadSize() int {
+	return r.c.PayloadSize()
+}
+
+// ReadWrite returns a Transport using map that can read or write with Get() and Put()
+func ReadWrite(c *Client, rwCap common.ReadWriteCap) RWClient {
+	m := new(rwMap)
+	m.c = c
+	m.rwCap = rwCap
+	return m
+}
+
+// ReadOnly returns a Transport using map that can read with Get() only
+func ReadOnly(c *Client, roCap common.ReadOnlyCap) ROClient {
+	m := new(roMap)
+	m.c = c
+	m.roCap = roCap
+	return m
+}
+
+// WriteOnly returns a Transport using map that can write with Put() only
+func WriteOnly(c *Client, woCap common.WriteOnlyCap) WOClient {
+	m := new(woMap)
+	m.c = c
+	m.woCap = woCap
+	return m
+}
+
+// create a duplex using a shared secret
+func DuplexFromSeed(c *Client, initiator bool, secret []byte) RWClient {
+	salt := []byte("duplex initialized from seed is not for multi-party use")
+	keymaterial := hkdf.New(hash, secret, salt, nil)
+	var err error
+	var pk1, pk2 *eddsa.PrivateKey
+	// return the listener or dialer side of caps from seed
+	if initiator {
+		if pk1, err = eddsa.NewKeypair(keymaterial); err != nil {
+			panic(err)
+		}
+		if pk2, err = eddsa.NewKeypair(keymaterial); err != nil {
+			panic(err)
+		}
+	} else {
+		if pk2, err = eddsa.NewKeypair(keymaterial); err != nil {
+			panic(err)
+		}
+		if pk1, err = eddsa.NewKeypair(keymaterial); err != nil {
+			panic(err)
+		}
+	}
+
+	// initialize root capabilities for both keys
+	rw1 := common.NewRWCap(pk1)
+	rw2 := common.NewRWCap(pk2)
+
+	// initiator socket
+	return Duplex(c, rw1.ReadOnly(), rw2.WriteOnly())
+}
+
+// duplex holds a pair of ROClient and WOClient and implements
+// RWClient with different read/write root capabilities so that a pair
+// of clients may use the capabilities to communicate unidirectionally
+type duplex struct {
+	ro ROClient // used to read data to client
+	wo WOClient // used to send data to peer
+}
+
+// Put implements RWClient.Put
+func (s *duplex) Put(addr []byte, payload []byte) error {
+	return s.wo.Put(addr, payload)
```

Comment:
here's another example of a method that needs to be rewritten so that we can test the math separately from the io